### PR TITLE
Calendar

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# Description
Add a new functionality in the menu bar called "Calendar" which allows users to select a timestamp (in UTC seconds). The viewer will then jump to the first event with the timestamp greater or equal to the selected one.
The UI looks like the following:

# Validation performed
1. Manually test the functionality.
2. Add unit tests for UTC timestamp converting.

### CODE REVIEW

**Comment by LinZhihao-723** [comment]
Here are two screenshots of the viewer:

1. Menu bar
bece782480627bc9f040a9876e4bb11

2. Calendar opened
9fee379d5dd0d1743a2119c376f1b22

File: src/Viewer/components/MenuBar/MenuBar.js [hunk 1]
```diff
@@ -135,6 +157,29 @@ export function MenuBar ({
         }
     };
 
+    const openCalendar = () => {
+        handleShowCalendar();
+    };
+
+    const closeCalendar = () => {
+        handleCloseCalendar();
+    };
```
Comment (Bill-hbrhbr) [inline src/Viewer/components/MenuBar/MenuBar.js:166]: These two functions seem a bit unnecessary

File: src/Viewer/services/ActionHandler.js [hunk 2]
```diff
@@ -121,6 +121,25 @@ class ActionHandler {
...
+    /**
+     * Goes to the specified log event with the specified timestamp.
+     * Go to new page if needed.
+     * @param {number} timestamp
+     */
+    changeEventWithTimestamp (timestamp) {
+        if (!isNumeric(timestamp)) {
+            throw (new Error("Invalid timestamp provided."));
+        }
+        this._logFile.state.logEventIdx = this._logFile.getLogEventIdxFromTimestamp(timestamp);
```
Comment (Bill-hbrhbr) [inline src/Viewer/services/ActionHandler.js:133]: I think the newest private class variable syntax is using #. I'm not sure if the _ prefix here means something should be private.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

File: src/Viewer/services/ActionHandler.js [hunk 3]
```diff
@@ -121,6 +121,25 @@ class ActionHandler {
         this._updateStateCallback(CLP_WORKER_PROTOCOL.UPDATE_STATE, this._logFile.state);
     };
 
+    /**
+     * Goes to the specified log event with the specified timestamp.
+     * Go to new page if needed.
+     * @param {number} timestamp
+     */
+    changeEventWithTimestamp (timestamp) {
+        if (!isNumeric(timestamp)) {
```
Comment (Bill-hbrhbr) [inline src/Viewer/services/ActionHandler.js:130]: Don't wanna be that guy, but

`false === isNumeric(timestamp)`

File: src/Viewer/components/MenuBar/MenuBar.js (hunk 1 above)
Comment (LinZhihao-723) [reply src/Viewer/components/MenuBar/MenuBar.js:166]: True, I agree

File: src/Viewer/components/MenuBar/MenuBar.js [hunk 4]
```diff
@@ -51,8 +67,14 @@ export function MenuBar ({
     const {theme, switchTheme} = useContext(ThemeContext);
 
     const [eventsPerPage, setEventsPerPage] = useState(logFileState.pages);
+    const [calendarDateTime, setCalendarDateTime] = useState(
+        new Date(Math.floor(Date.now() / 1000) * 1000)
```
Comment (Bill-hbrhbr) [inline src/Viewer/components/MenuBar/MenuBar.js:71]: Could make the millisecond truncation into a function.

File: src/Viewer/components/MenuBar/MenuBar.js [hunk 5]
```diff
@@ -135,6 +157,29 @@ export function MenuBar ({
...
+        handleShowCalendar();
+    };
+
+    const closeCalendar = () => {
+        handleCloseCalendar();
+    };
+
+    const handleCalendarChange = (newDate) => {
+        if (newDate instanceof Date) {
+            setCalendarDateTime(newDate);
```
Comment (Bill-hbrhbr) [inline src/Viewer/components/MenuBar/MenuBar.js:170]: use the millisecond truncation function here as well for consistent calendarDateTime state behavior (although maybe you already pass in a date value with the millisecond part removed).

**Review by Bill-hbrhbr — CHANGES_REQUESTED** [review]
Leaving the rest to our protagonist Kirk,

**Comment by kirkrodrigues** [comment]
Since the UI has been rewritten, can we close this PR and open a feature request for this feature?

**Comment by junhaoliao** [comment]
The feature is now tracked by #117 & #118 . Closing this since we will have to open separate PRs when the code is ready to be submitted.

**Pull request closed without merging**